triggers/daml/Daml/Trigger/Internal.daml

```diff
@@ -49,16 +51,45 @@ data ACS = ACS
 -- | TriggerA is the type used in the `rule` of a DAML trigger.
 -- Its main feature is that you can call `emitCommands` to
 -- send commands to the ledger.
-newtype TriggerA a = TriggerA (TriggerRule TriggerAState a)
-  deriving (Functor, Applicative, Action, HasTime)
+newtype TriggerA a = TriggerA { runTriggerA : ACS -> TriggerRule TriggerAState a }
+
+instance Functor TriggerA where
+  fmap f (TriggerA r) = TriggerA $ rliftFmap fmap f r
+
+instance Applicative TriggerA where
+  pure = TriggerA . rliftPure pure
+  TriggerA ff <*> TriggerA fa = TriggerA $ rliftAp (<*>) ff fa
+
+instance Action TriggerA where
+  TriggerA fa >>= f = TriggerA $ rliftBind (>>=) fa (runTriggerA . f)
+
+instance HasTime TriggerA where
+  getTime = TriggerA $ const getTime
 
 -- | TriggerStateA is the type used in the `updateState` of a DAML
 -- trigger.  It has similar actions in common with `TriggerA`, but
 -- cannot use `emitCommands` or `getTime`.
-newtype TriggerStateA s a = TriggerStateA { runTriggerStateA : State s a }
-  deriving (Functor, Applicative, Action)
+newtype TriggerStateA s a = TriggerStateA { runTriggerStateA : ACS -> State s a }
+
+instance Functor (TriggerStateA s) where
+  fmap f (TriggerStateA r) = TriggerStateA $ rliftFmap fmap f r
+
+instance Applicative (TriggerStateA s) where
+  pure = TriggerStateA . rliftPure pure
+  TriggerStateA ff <*> TriggerStateA fa = TriggerStateA $ rliftAp (<*>) ff fa
 
-deriving instance ActionState s (TriggerStateA s)
+instance Action (TriggerStateA s) where
+  TriggerStateA fa >>= f = TriggerStateA $ rliftBind (>>=) fa (runTriggerStateA . f)
+
+instance ActionState s (TriggerStateA s) where
+  get = TriggerStateA $ const get
+  put = TriggerStateA . const . put
+  modify = TriggerStateA . const . modify
+
+-- | Features possible in both `updateState` and `rule`.
+class ActionTriggerUpdate m where
+  -- | Extract the contracts of a given template from the ACS.
+  query : Template a => m [(ContractId a, a)]
```

Comment:
Exactly like in `Script`, but the trigger's `initialize` still needs to use `getContracts` on its explicit `ACS` argument with the prior signature. Thoughts @cocreature ?

Reply:
Not quite sure about this. It feels a bit weird to have `query` for one but `getContracts` for the other. What are your thoughts into turning this into a reader-like monad? 3 Monads are arguably a bit confusing but 2 monads and one function which requires a different API doesn’t seem much better. @aherrmann-da What are your thoughts on this?

Reply:
I'd tend towards three monads rather than two plus one function. Having that kind of difference between `initialize` and `updateState` seems a bit awkward if I image an example where the trigger state depends on the ACS.

Reply:
The affinity between `updateState` and `rule` is more obvious to me than that with `initialize`, but it is certainly a bit awkward without the third monad. Do you think it should be in this PR too?

Reply:
I think this PR makes sense to avoid introducing an inconsistency between `update` and `initialize`.